experimental/AlgebraicStatistics/src/CI.jl

```diff
@@ -0,0 +1,225 @@
+
+struct CIStmt
+  I::Vector{String}
+  J::Vector{String}
+  K::Vector{String}
+end
+
+@doc raw"""
+    ci_stmt(I, J, K; symmetric=true, semigraphoid=true)
+
+A conditional independence statement asserting that `I` is independent
+of `J` given `K`. These parameters are lists of names of random variables.
+The sets `I` and `J` must be disjoint as this package cannot yet deal
+with functional dependencies.
+
+If `symmetric` is `true`, CI statements are assumed to be symmetric in
+their `I` and `J` components. The constructor then reorders the arguments
+to make the `I` field lexicographically smaller than the `J` to ensure
+that comparisons and hashing respect the symmetry.
+
+If `semigraphoid` is set to `true`, the constructor also removes elements
+in the intersection of `I` and `K` from `I` (and symetrically removes the
+intersection of `J` and `K` from `J`).
+
+As all three fields are sets, each of them may be deduplicated and sorted
+to ensure consistent comparison and hashing.
+
+## Examples
+
+``` jldoctest
+julia> ci_stmt(["A"], ["B"], ["X"])
+[A ⫫ B | X]
+
+julia> ci_stmt(["1"], ["2", "3"], ["4", "5"])
+[1 ⫫ {2, 3} | {4, 5}]
+```
+"""
+function ci_stmt(I, J, K; symmetric=true, semigraphoid=true)
+  if length(intersect(I, J)) > 0
+    error("Functional dependence statements are not yet implemented")
+  end
+  if symmetric && I > J
+    I, J = J, I
+  end
+  if semigraphoid
+    I = setdiff(I, K)
+    J = setdiff(J, K)
+  end
+  CIStmt(sort(unique(I)), sort(unique(J)), sort(unique(K)))
+end
+
+@doc raw"""
+    Base.:(==)(lhs::CIStmt, rhs::CIStmt)
+
+Compares `CIStmt`s for identity in all their three fields.
+"""
+Base.:(==)(lhs::CIStmt, rhs::CIStmt) =
+  lhs.I == rhs.I && lhs.J == rhs.J && lhs.K == rhs.K
+
+@doc raw"""
+    Base.hash(stmt:;CIStmt, h::UInt)
+
+Computes the hash of a `CIStmt`.
+"""
+Base.hash(stmt::CIStmt, h::UInt) =
+  foldr(hash, stmt.I, stmt.J, stmt.K; init=hash(CIStmt, h))
+
+@doc raw"""
+    CI"I...,J...|K..."
+
+A literal syntax for denoting CI statements is provided for cases in which
+all variable names consist of a single character. If `I` and `J` only consist
+of a single element, then even the comma may be omitted. Once the three sets
+are extracted, `ci_stmt` is called.
+
+## Examples
+
+``` jldoctest
+julia> CI"AB|X"
+[A ⫫ B | X]
+
+julia> CI"1,23|5424"
+[1 ⫫ 3 | {2, 4, 5}]
+```
+"""
+macro CI_str(str)
+  # General syntax "12,34|567"
+  m = match(r"^(.+),(.+)[|](.*)$", str)
+  # Short syntax for elementary statements "12|345"
+  if m == nothing
+    m = match(r"^(.)(.)[|](.*)$", str)
+  end
+  # Give up
+  if m == nothing
+    throw(ArgumentError(str * " is not a CI statement"))
+  end
+  parse_arg(s) = unique([string(c) for c in s])
+  I, J, K = map(s -> parse_arg(s), m)
+  return ci_stmt(I, J, K)
+end
+
+function Base.show(io::IO, stmt::CIStmt)
+  fmt(K) = length(K) == 1 ? string(K[1]) : "{" * join([string(x) for x in K], ", ") * "}"
+  print(io, "[$(fmt(stmt.I)) ⫫ $(fmt(stmt.J)) | $(fmt(stmt.K))]")
```

Comment:
Unconditional use of unicode here goes against our conventions, see https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/#Unicode and https://docs.oscar-system.org/dev/DeveloperDocumentation/printing_details/#Unicode-printing for details

Reply:
Thanks for the reminder. I changed that to conditional.